src/message_bus.zig

```diff
@@ -60,73 +233,419 @@ pub const MessageBus = struct {
+        const server_addr = bus.configuration[bus.server.replica];
+        self.fd = os.socket(server_addr.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0) catch return;
+
+        self.peer = .{ .replica = replica };
+        bus.replicas[replica] = self;
+        bus.replicas_connected += 1;
+
+        self.recv_in_progress = true;
+        bus.io.connect(
+            *Connection,
+            self,
+            on_connect,
+            // We need to use the recv_completion here
+            &self.recv_completion,
+            self.fd,
+            // TODO: is this actually safe to pass by value?
+            // Should IO just take an std.net.Address or a pointer?
+            bus.configuration[replica].any,
+            bus.configuration[replica].getOsSockLen(),
+        );
+    }
+
+    fn on_connect(self: *Connection, completion: *IO.Completion, result: IO.ConnectError!void) void {
+        self.recv_in_progress = false;
+        result catch |err| {
+            log.err("error connecting to {}: {}", .{ self.peer, err });
+            self.shutdown();
+            return;
+        };
+        log.debug("connected to {}", .{self.peer});
+        self.recv_header();
+    }
+
+    /// Given a newly accepted fd, start receiving messages on it.
+    /// Callbacks will be continuously re-registered until shutdown() is
+    /// called and the connection is terminated.
+    pub fn on_accept(self: *Connection, fd: os.socket_t) void {
+        assert(self.peer == .none);
+        assert(self.fd == -1);
+        self.fd = fd;
+        self.peer = .unknown;
+        self.message_bus.connections_in_use += 1;
+        self.recv_header();
+    }
+
+    /// Add a message to the connection's send queue, starting a send operation
+    /// if the queue was previously empty.
+    pub fn send_message(self: *Connection, envelope: *Envelope) void {
+        assert(self.peer == .client or self.peer == .replica);
+        const queue_was_empty = self.send_queue.out == null;
+        self.send_queue.push(envelope);
+        // If the queue was not empty, the message will be sent after the
+        // messages currently being sent.
+        if (queue_was_empty) self.send();
+    }
+
+    /// Clean up an active connection and reset it to its initial, unusued, state.
+    /// This reset does not happen instantly as currently in progress operations
+    /// must first be stopped.
+    pub fn shutdown(self: *Connection) void {
+        assert(self.peer != .shutting_down and self.peer != .none);
+        assert(self.fd != -1);
+        // The shutdown syscall will cause currently in progress send/recv
+        // operations to be terminated with an error while keeping the fd open.
+        const rc = os.linux.shutdown(self.fd, os.SHUT_RDWR);
+        switch (os.errno(rc)) {
+            0 => {},
+            os.EBADF => unreachable,
+            os.EINVAL => unreachable,
+            os.ENOTCONN => @panic("TODO: is this possible?"),
```

Comment:
Good catch! Asserting that there is no send or recv pending and continuing on with the shutdown logic sounds like the right way to handle this. The only complication is that we currently do not distinguish between a recv or a connect operation being in progress using the `recv_completion` but this can of course be remedied.

Reply:
Great, we could then use our new `.connecting` status to distinguish between a recv or a connect operation?